server/lib/chat/chat-service.test.ts

```diff
@@ -0,0 +1,1505 @@
+      addUserToChannelMock.mockResolvedValue(user1TestChannelEntry)
+      addMessageToChannelMock.mockResolvedValue(joinUser1TestChannelMessage)
+      createChannelMock.mockResolvedValue(testChannel)
+
+      await chatService.joinChannel(testChannel.name, user1.id)
+
+      expect(addUserToChannelMock).toHaveBeenCalledWith(user1.id, testChannel.id, dbClient)
+      expect(addMessageToChannelMock).toHaveBeenCalledWith(
+        user1.id,
+        testChannel.id,
+        { type: ServerChatMessageType.JoinChannel },
+        dbClient,
+      )
+
+      // NOTE(2Pac): Since there are no users in this channel, we don't bother checking to see if
+      // the publish message was sent/received.
+
+      expect(nydus.subscribeClient).toHaveBeenCalledWith(client1, getChannelPath(testChannel.id), {
+        action: 'init3',
+        channelInfo: testChannel,
+        activeUserIds: [user1.id],
+        selfPermissions: channelPermissions,
+      })
+      expect(nydus.subscribeClient).toHaveBeenCalledWith(
+        client1,
+        getChannelUserPath(testChannel.id, user1.id),
+        undefined,
+      )
+    })
+  })
+
+  describe('leaveChannel', () => {
+    const removeUserFromChannelMock = asMockedFunction(removeUserFromChannel)
+
+    beforeEach(async () => {
+      removeUserFromChannelMock.mockResolvedValue({ newOwnerId: undefined })
+    })
+
+    test('should throw if not in channel', async () => {
+      await expect(
+        chatService.leaveChannel(testChannel.id, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"Must be in channel to leave it"`)
+    })
+
+    describe('when in ShieldBattery channel', () => {
+      beforeEach(async () => {
+        await joinUserToChannel(
+          user1,
+          shieldBatteryChannel,
+          user1ShieldBatteryChannelEntry,
+          joinUser1ShieldBatteryChannelMessage,
+        )
+        await joinUserToChannel(
+          user2,
+          shieldBatteryChannel,
+          user2ShieldBatteryChannelEntry,
+          joinUser2ShieldBatteryChannelMessage,
+        )
+      })
+
+      test("should throw if it's not allowed", async () => {
+        flagsMock.CAN_LEAVE_SHIELDBATTERY_CHANNEL = false
+
+        await expect(
+          chatService.leaveChannel(shieldBatteryChannel.id, user1.id),
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`"Can't leave ShieldBattery channel"`)
+      })
+
+      test("works when it's allowed", async () => {
+        flagsMock.CAN_LEAVE_SHIELDBATTERY_CHANNEL = true
+
+        await chatService.leaveChannel(shieldBatteryChannel.id, user1.id)
+
+        expect(removeUserFromChannelMock).toHaveBeenCalledWith(user1.id, shieldBatteryChannel.id)
+        expect(client2.publish).toHaveBeenCalledWith(getChannelPath(shieldBatteryChannel.id), {
+          action: 'leave2',
+          userId: user1.id,
+          newOwnerId: undefined,
+        })
+        expect(nydus.unsubscribeClient).toHaveBeenCalledWith(
```

Comment:
Would prefer triggering something that publishes to the channel (like sending a message) and checking that the client did not receive it

Reply:
Still too much coupling to the actual mechanics versus testing that the state is correct. Ideally our tests wouldn't need to know anything about the paths being used at all, but unfortunately they're significant to the behavior (and not really avoidable unless you hooked up the client code and examined the reducer state or something)